packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-engine.ts

```diff
@@ -0,0 +1,888 @@
+          const runOrdersConsumed = [0];
+
+          for (const node of tranche) {
+            const factory = this.actionFromNode(node);
+
+            const nodeType = this.nodeTypeFromNode(node);
+
+            const result = factory.produce({
+              actionName: actionName(node, sharedParent),
+              runOrder,
+              stage: pipelineStage,
+              artifacts: this.artifacts,
+              scope: obtainScope(this.pipeline, stageName),
+              fallbackArtifact: this._fallbackArtifact,
+              queries: structure.queries,
+              // If this step happens to produce a CodeBuild job, set the default options
+              codeBuildDefaults: nodeType ? this.codeBuildDefaultsFor(nodeType) : undefined,
+              beforeSelfMutation,
+            });
+
+            if (node.data?.type === 'self-update') {
+              beforeSelfMutation = false;
+            }
+
+            this.postProcessNode(node, result);
+
+            runOrdersConsumed.push(result.runOrdersConsumed);
+          }
+
+          runOrder += Math.max(...runOrdersConsumed);
+        }
+      }
+    }
+  }
+
+  /**
+   * Do additional things after the action got added to the pipeline
+   *
+   * Some minor state manipulation of CodeBuild projects and pipeline
+   * artifacts.
+   */
+  private postProcessNode(node: AGraphNode, result: CodePipelineActionFactoryResult) {
+    const nodeType = this.nodeTypeFromNode(node);
+
+    if (result.project) {
+      const dockerUsage = dockerUsageFromCodeBuild(nodeType ?? CodeBuildProjectType.STEP);
+      if (dockerUsage) {
+        for (const c of this.dockerCredentials) {
+          c.grantRead(result.project, dockerUsage);
+        }
+      }
+
+      if (nodeType === CodeBuildProjectType.SYNTH) {
+        this._synthProject = result.project;
+      }
+    }
+
+    if (node.data?.type === 'step' && node.data.step.primaryOutput?.primaryOutput && !this._fallbackArtifact) {
+      this._fallbackArtifact = this.artifacts.toCodePipeline(node.data.step.primaryOutput?.primaryOutput);
+    }
+  }
+
+  /**
+   * Make an action from the given node and/or step
+   */
+  private actionFromNode(node: AGraphNode): ICodePipelineActionFactory {
+    switch (node.data?.type) {
+      // Nothing for these, they are groupings (shouldn't even have popped up here)
+      case 'group':
+      case 'stack-group':
+      case undefined:
+        throw new Error(`makeAction: did not expect to get group nodes: ${node.data?.type}`);
+
+      case 'self-update':
+        return this.selfMutateAction();
+
+      case 'publish-assets':
+        return this.publishAssetsAction(node, node.data.assets);
+
+      case 'prepare':
+        return this.createChangeSetAction(node.data.stack);
+
+      case 'execute':
+        return this.executeChangeSetAction(node.data.stack, node.data.captureOutputs);
+
+      case 'step':
+        return this.actionFromStep(node, node.data.step);
+    }
```

Comment:
Add "default" and throw

Reply:
TypeScript will produce a compiler error if the switch isn't exhaustive.